### PR TITLE
Image tags

### PR DESCRIPTION
**Pull request by erikperkins**
Fixes issue #10.

Add tags to images and display tags along with the images.

### CODE REVIEW

File: app/views/images/new.html.erb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <%= form_for @image do |f| %>
   <%= f.label :url %>
   <%= f.text_field :url %>
+  <%= f.text_field :tag_list %>
```
Comment (shfung) [inline app/views/images/new.html.erb]: Do we need label for this field?

File: config/initializers/acts_as_taggable_patch.rb [hunk 2]
```diff
@@ -0,0 +1 @@
+ActsAsTaggableOn.force_binary_collation = true
```
Comment (shfung) [inline config/initializers/acts_as_taggable_patch.rb]: I'm wondering why do we need this setting?

File: db/migrate/20160223173235_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# This migration comes from acts_as_taggable_on_engine (originally 1)
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  def self.up
+    create_table :tags do |t|
+      t.string :name
```
Comment (shfung) [inline db/migrate/20160223173235_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: I understand that these migration and table structures are provided by the gem, but I'm wondering wether we need to add some not-null constraints to some of the columns?

File: db/migrate/20160223173239_change_collation_for_tag_names.acts_as_taggable_on_engine.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# This migration comes from acts_as_taggable_on_engine (originally 5)
+# This migration is added to circumvent issue #623 and have special characters
+# work properly
+class ChangeCollationForTagNames < ActiveRecord::Migration
+  def up
+    if ActsAsTaggableOn::Utils.using_mysql?
```
Comment (shfung) [inline db/migrate/20160223173239_change_collation_for_tag_names.acts_as_taggable_on_engine.rb]: Do we need this migration?

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -3,17 +3,22 @@
...
-      assert_response :success
-    end
+  test 'create image without url' do
+    post :create, { image: { url: '', tag_list: 'haskell, programming' } }
+    assert_response :unprocessable_entity
+  end
+
+  test 'create image without tags' do
+    post :create, { image: { url: '', tag_list: '' } }
+    assert_response :unprocessable_entity
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: It seems that the content of this test doesn't match entirely with its name. The `unprocessable_entity` seems to be triggered by the empty url instead of the empty tag_lists.
Maybe we can add one more test with valid url but empty tag_list as params.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -6,16 +6,23 @@ def new
 
 
   def create
-    image = params[:image]
+    image = image_params
 
     @image = Image.new url: image[:url]
+    @image.tag_list.add image[:tag_list], parse: true
```
Comment (shfung) [inline app/controllers/images_controller.rb]: It seems that the `acts-as-taggable-on` gem allow any string to be the tag content.  Maybe we can consider sanitizing the string to avoid any potential SQL injection or XSS.

File: test/models/image_test.rb [hunk 7]
```diff
@@ -12,4 +12,10 @@ class ImageTest < ActiveSupport::TestCase
     assert_not Image.new.save
   end
 
+  test 'add image tags' do
+    url = 'http://imgs.xkcd.com/comics/haskell.png'
+    tags = ['haskell', 'programming']
+    assert_predicate Image.new(url: url, tag_list: tags), :valid?
+  end
```
Comment (rmacklin) [inline test/models/image_test.rb]: I'm not sure of the value of this test as it's currently written.

File: test/controllers/images_controller_test.rb (hunk 5 above)
Comment (bboe) [inline test/controllers/images_controller_test.rb]: I think it's okay if we have only one invalid test in the controller test since we rely on testing the effectiveness of each validator in the model test.

On a separate note, in such a failure, what else should we test for?

File: app/controllers/images_controller.rb (hunk 6 above)
Comment (bboe) [inline app/controllers/images_controller.rb]: @shfung I'm not sure if I agree that sanitizing things going into our database is the right approach. Sanitizing on display makes things much less complicated.

We should also be able to rely on `acts_as_taggable_on` to properly handle SQL injection. Of course we should verify that.

File: db/migrate/20160223173235_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 3 above)
Comment (bboe) [inline db/migrate/20160223173235_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: > I'm wondering wether we need to add some not-null constraints to some of the columns

:+1:

File: app/controllers/images_controller.rb (hunk 6 above)
Comment (shfung) [inline app/controllers/images_controller.rb]: Fair enough.

On a similar note: is there any format restriction or length limitation that we want to enforce on the tags?

Comment (bboe) [inline app/controllers/images_controller.rb]: Each tag is a string in the rails migration. So whatever the database limit is on that field. I don't believe we explicitly test for field boundaries unless we have a reason to.

 Each image can have an any number of tags. A image with tens+ tags could be problematic, but that's not really something we're concerned with at this point (MVP).

File: config/initializers/acts_as_taggable_patch.rb (hunk 2 above)
Comment (erikperkins) [inline config/initializers/acts_as_taggable_patch.rb]: It's the suggested workaround https://github.com/mbleigh/acts-as-taggable-on#for-mysql-users for a bug in acts_as_taggable_on https://github.com/mbleigh/acts-as-taggable-on/issues/623. Heroku doesn't use MySQL, of course, but I included it anyway to be safe.

File: db/migrate/20160223173239_change_collation_for_tag_names.acts_as_taggable_on_engine.rb (hunk 4 above)
Comment (erikperkins) [inline db/migrate/20160223173239_change_collation_for_tag_names.acts_as_taggable_on_engine.rb]: Given the current deployment environment, it's not necessary, but it's included for the same reason as the line in config/initializers/acts_as_taggable_patch.rb.

File: app/controllers/images_controller.rb (hunk 6 above)
Comment (erikperkins) [inline app/controllers/images_controller.rb]: I'll leave this one alone for now.

File: db/migrate/20160223173235_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb (hunk 3 above)
Comment (erikperkins) [inline db/migrate/20160223173235_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb]: Addressed in dc45ed376e37a0667cbd53cb452ba06d429bd005

File: app/views/images/new.html.erb (hunk 1 above)
Comment (erikperkins) [inline app/views/images/new.html.erb]: Addressed in 1e3fb9f7996ab7f3fc7aabe7243ade0d69f05525

File: test/controllers/images_controller_test.rb (hunk 5 above)
Comment (erikperkins) [inline test/controllers/images_controller_test.rb]: Addressed in 03251743579f92c783e2f3aa43e0bb8b5e98fe40

I'm not sure what else we should test for, since the specification doesn't require images to have tags yet. The only way an image can be invalid (in this branch) is if it is submitted with a blank URL.

File: test/models/image_test.rb (hunk 7 above)
Comment (erikperkins) [inline test/models/image_test.rb]: Addressed in 03251743579f92c783e2f3aa43e0bb8b5e98fe40

**Comment by erikperkins** [comment]
@shfung @rmacklin @bboe How does it look now?

**Comment by rmacklin** [comment]
> How does it look now?

It looks like appropriate squashing has not been done yet. Once that's done I can take another look.

**Comment by erikperkins** [comment]
@rmacklin  Well, rebasing wiped out a bunch of your inline comments and my replies, but I rearranged things into more appropriate commits.

**Comment by rmacklin** [comment]
@erikperkins Looks really good. I see that [these tests](https://github.com/AppFolioOnboarding/perkins_eidolon/blob/9a1a42ec398fe196b4998e500e40060ea884e5dc/test/controllers/images_controller_test.rb#L52-L60) made it back even though it had been mentioned that it could be overtesting to test the different validation errors at the controller level. But at this point, it's not a big deal so if you wanna leave it I don't mind.

**Comment by erikperkins** [comment]
@rmacklin Okay, I changed the name of the test in [https://github.com/AppFolioOnboarding/perkins_eidolon/blob/7420a969090f665ff35e2025a7d9714201b3ee1a/test/models/image_test.rb, image_test.rb](url) and removed the two spurious tests in [https://github.com/AppFolioOnboarding/perkins_eidolon/blob/9a1a42ec398fe196b4998e500e40060ea884e5dc/test/controllers/images_controller_test.rb#L52-L60, image_controller_test.rb](url)

**Comment by rmacklin** [comment]
:+1:

File: db/schema.rb [hunk 8]
```diff
@@ -19,4 +19,24 @@
     t.datetime "updated_at", null: false
   end
 
+  create_table "taggings", force: :cascade do |t|
```
Comment (bboe) [inline db/schema.rb]: Which of these fields can be set `null: False`? Ideally we'd like only fields that need `null: true` to allow null values.
